File: Programmers/Level_2/스킬트리.py
'''
https://programmers.co.kr/learn/courses/30/lessons/49993

skill_trees에서 단어 하나 뽑아옴
단어안에 글자가 skill에 포함된다면
skill 순서에 맞는지 확인하기 위해 idx 변수로 관리

글자가 skill에 포함되는데 순서에 어긋난다면 오답처리
순서에 맞다면 idx 증가, 다음 글자가 다음 skill 순서와 비교되도록 함 
'''

def solution(skill, skill_trees):
    answer = 0
    
    for st in skill_trees:
        idx = 0
        answer += 1

        for i in range(len(st)):
            if st[i] in skill:
                if st[i] != skill[idx]:
                    answer -= 1
                    break
                else:
                    idx += 1

    return answer

File: Programmers/Level_2/Swift_Code/Swift_Code/스킬트리.swift
//
//  스킬트리.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/03/24.
//

/*
 시간복잡도에서 우위: solution1
 Swift 스러움: solution2
 */

import Foundation

func solution1(_ skill: String, _ skill_trees: [String]) -> Int {
    var answer = 0
    let s = Array(skill)

    for skill_tree in skill_trees {
        let st = Array(skill_tree)
        var idx = 0
        answer += 1

        for i in 0..<st.count {
            if s.contains(st[i]) {
                if st[i] != s[idx] {
                    answer -= 1
                    break
                } else {
                    idx += 1
                }
            }
        }
    }

    return answer
}

func findStartWith(_ skill: String, _ skillTree: String) -> Bool {
    let skillInSkillTree = skillTree.filter { skill.contains($0) }
    return skill.starts(with: skillInSkillTree)
}

func solution2(_ skill:String, _ skill_trees:[String]) -> Int {
    skill_trees
        .map { findStartWith(skill, $0) }
        .filter { $0 }
        .count
}

// print(solution2("CBD", ["BACDE", "CBADF", "AECB", "BDA"]))

